android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Router.kt

```diff
@@ -9,50 +9,58 @@ import com.badoo.mvicore.binder.Binder
 import com.badoo.ribs.core.routing.configuration.Transaction.MultiConfigurationCommand.SaveInstanceState
 import com.badoo.ribs.core.routing.configuration.Transaction.MultiConfigurationCommand.Sleep
 import com.badoo.ribs.core.routing.configuration.Transaction.MultiConfigurationCommand.WakeUp
-import com.badoo.ribs.core.routing.configuration.feature.BackStackFeature
 import com.badoo.ribs.core.routing.configuration.feature.ConfigurationFeature
-import com.badoo.ribs.core.routing.configuration.feature.operation.BackStackOperation
-import com.badoo.ribs.core.routing.configuration.feature.operation.Pop
-import com.badoo.ribs.core.routing.configuration.feature.operation.canPop
-import com.badoo.ribs.core.routing.configuration.feature.operation.canPopOverlay
 import com.badoo.ribs.core.routing.configuration.toCommands
+import com.badoo.ribs.core.routing.history.Routing
 import com.badoo.ribs.core.routing.transition.handler.TransitionHandler
-import com.badoo.ribs.core.view.RibView
+import com.jakewharton.rxrelay2.PublishRelay
+import com.jakewharton.rxrelay2.Relay
+import io.reactivex.ObservableSource
+import io.reactivex.Observer
+import io.reactivex.disposables.CompositeDisposable
 
-abstract class Router<C : Parcelable, Permanent : C, Content : C, Overlay : C, V : RibView>(
+abstract class Router<C : Parcelable>(
     buildParams: BuildParams<*>,
-    private val initialConfiguration: Content,
-    private val permanentParts: List<Permanent> = emptyList(),
-    private val transitionHandler: TransitionHandler<C>? = null
-) : NodeAware,
+    protected val routingSource: RoutingSource<C>,
+    private val transitionHandler: TransitionHandler<C>? = null,
+    private val clientChildActivator: ChildActivator<C> = UnhandledChildActivator()
+) : ConfigurationResolver<C>,
+    NodeAware,
     NodeLifecycleAware,
     ViewLifecycleAware,
     SavesInstanceState,
-    SubtreeBackPressHandler,
-    ConfigurationResolver<C> {
-
-    companion object {
-        internal const val BUNDLE_KEY = "Router"
+    SubtreeBackPressHandler by routingSource,
+    ObservableSource<Router.Event<C>> {
+
+    @ExperimentalApi
+    sealed class Event<C : Parcelable> {
+        data class Activated<C : Parcelable>(
+            val routing: Routing<C>,
+            val nodes: List<Node<*>>
+        ) : Event<C>()
+
+        data class Deactivated<C : Parcelable>(
+            val routing: Routing<C>,
+            val nodes: List<Node<*>>
+        ) : Event<C>()
     }
 
     private val binder = Binder()
-    private val savedInstanceState = buildParams.savedInstanceState?.getBundle(BUNDLE_KEY)
-    private val timeCapsule: AndroidTimeCapsule = AndroidTimeCapsule(this.savedInstanceState)
-    private lateinit var backStackFeature: BackStackFeature<C>
+    private val disposables = CompositeDisposable()
+    private val timeCapsule: AndroidTimeCapsule = AndroidTimeCapsule(buildParams.savedInstanceState)
+    private val events: Relay<Event<C>> = PublishRelay.create()
```

Comment:
where you accept any events?

Reply:
fyi 
you can move it to constructor and add delegate it to ObservableSource.

Reply:
Good catch, this is a leftover and should be removed